core/src/main/scala/com/salesforce/op/stages/impl/selector/ModelSelector.scala

```diff
@@ -138,16 +137,19 @@ E <: Estimator[_] with OpPipelineStage2[RealNN, OPVector, Prediction]]
 
     implicit val spark = dataset.sparkSession
 
-    val datasetWithID =
+    val datasetWithIDPre =
       if (dataset.columns.contains(DataFrameFieldNames.KeyFieldName)) {
         dataset.select(in1.name, in2.name, DataFrameFieldNames.KeyFieldName)
       } else {
         dataset.select(in1.name, in2.name)
           .withColumn(ModelSelectorNames.idColName, monotonically_increasing_id())
       }
-    require(!datasetWithID.isEmpty, "Dataset cannot be empty")
+    require(!datasetWithIDPre.isEmpty, "Dataset cannot be empty")
 
-    val splitterSummary = splitter.flatMap(_.preValidationPrepare(datasetWithID))
+    val splitterSummary = splitter.flatMap(_.preValidationPrepare(datasetWithIDPre))
+    val datasetWithID = splitter.collect {
+      case dc: DataCutter => dc.validationPrepare(datasetWithIDPre)
+    }.getOrElse(datasetWithIDPre)
```

Comment:
if you want to avoid the match on case for the data prepare we could modify the return type of  preValidationPrepare to optionally return a dataset. then the data cutter preValidationPrepare could cut the data and the validationPrepare could just return the dataset unchanged...

Reply:
yes. hacked something together to see if it's generally a good approach